File: programmers/pg_lv3_네트워크.js
function solution(n, computers) {
  const graph = Array.from({ length: n }, () => []);
  for (let i = 0; i < n; i++) {
    for (let j = 0; j < n; j++) {
      if (i == j) continue;
      if (computers[i][j] == 1) graph[i].push(j);
    }
  }
  let count = 0;
  const visited = Array.from({ length: n }, () => false);

  const dfs = (k) => {
    visited[k] = true;

    graph[k].forEach((ele) => {
      if (!visited[ele]) dfs(ele);
    });
  };

  for (let k = 0; k < n; k++) {
    if (!visited[k]) {
      dfs(k);
      count++;
    }
  }

  return count;
}

File: python/programmers/pg_lv3_네트워크.py
def solution(n, computers):
    n = len(computers)
    count = 0
    graph = []
    for (idx1, connection) in enumerate(computers):
        tmpList = []
        for (idx2, i) in enumerate(connection):
            if idx1 == idx2: continue
            elif i == 1: tmpList.append(idx2)
        graph.append(tmpList)
        
    visited = [False for _ in range(n)]
    def dfs(idx):
        visited[idx] = True
        for i in graph[idx]:
            if not visited[i]: dfs(i)

    for idx in range(n):
        if not visited[idx]:
            count += 1
            dfs(idx)
    
    return count